File: leetcode/39.CombinationSum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);
        Stack<Integer> stack = new Stack<>();
        List<List<Integer>> ans = new ArrayList<>();
        
        this.process(candidates, target, 0, stack, ans);
        return ans;
    }
    
    private void process(int[] candidates, int target, int idx,
                        Stack<Integer> stack, List<List<Integer>> ans) {
        for (int i = idx; i < candidates.length; i++) {
            target -= candidates[i];
            if (target < 0) break;
            
            stack.push(candidates[i]);
            if (target == 0) {
                ans.add(new ArrayList(stack));
            } else {
                this.process(candidates, target, i, stack, ans);
            }
            
            target += candidates[i];
            stack.pop();
        }
    }
}

File: leetcode/39.combination-sum.py
#
# @lc app=leetcode id=39 lang=python3
#
# [39] Combination Sum
#
from typing import List

# @lc code=start
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates = sorted(candidates)
        ans, st = [], []
        self.process(candidates, target, 0, st, ans)
        return ans

    def process(
        self, candidates: List[int], 
        target: int,
        idx: int,
        st: List[int],
        ans: List[List[int]]
    ):
        for i in range(idx, len(candidates)):
            target -= candidates[i]
            if target < 0:
                break
            
            st.append(candidates[i])
            if target == 0:
                ans.append(st.copy())
            elif target > 0:
                self.process(candidates, target, i, st, ans)
            target += candidates[i]
            st.pop()


if __name__ == '__main__':
    candidates = [2, 3, 6, 7]
    # candidates = [2, 3, 5]
    ans = Solution().combinationSum(candidates=candidates, target=7)
    print(ans)

# @lc code=end

